cli/run.go

```diff
@@ -0,0 +1,139 @@
+	"This is useful when you just want to use one command to start a container. " +
+	"Also, you can specify --rm to delete it after the container quits. "
+
+// RunCommand use to implement 'run' command, it creates and starts a container.
+type RunCommand struct {
+	baseCommand
+	container
+	attach bool
+	stdin  bool
+	rm     bool
+}
+
+// Init initialize run command.
+func (rc *RunCommand) Init(c *Cli) {
+	rc.cli = c
+	rc.cmd = &cobra.Command{
+		Use:   "run a container",
+		Short: "Create a new container and start it",
+		Long:  runDescription,
+		Args:  cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return rc.runRun(args)
+		},
+		Example: runExample(),
+	}
+	rc.addFlags()
+}
+
+// addFlags adds flags for specific command.
+func (rc *RunCommand) addFlags() {
+	flagSet := rc.cmd.Flags()
+	flagSet.StringVar(&rc.name, "name", "", "Specify name of container")
+	flagSet.BoolVarP(&rc.tty, "tty", "t", false, "Allocate a tty device")
+	flagSet.StringSliceVarP(&rc.volume, "volume", "v", nil, "Bind mount volumes to container")
+	flagSet.StringVar(&rc.runtime, "runtime", "", "Specify oci runtime")
+	flagSet.BoolVarP(&rc.attach, "attach", "a", false, "Attach container's STDOUT and STDERR")
+	flagSet.BoolVarP(&rc.stdin, "interactive", "i", false, "Attach container's STDIN")
+	flagSet.BoolVarP(&rc.rm, "rm", "", false, "Remove the container after running")
+}
+
+// runRun is the entry of run command.
+func (rc *RunCommand) runRun(args []string) error {
+	config := rc.config()
+
+	ref, err := reference.Parse(args[0])
+	if err != nil {
+		return fmt.Errorf("failed to run container: %v", err)
+	}
+
+	config.Image = ref.String()
+	if len(args) == 2 {
+		config.Cmd = strings.Fields(args[1])
+	}
+	containerName := rc.name
+
+	apiClient := rc.cli.Client()
+	result, err := apiClient.ContainerCreate(config.ContainerConfig, config.HostConfig, containerName)
+	if err != nil {
+		return fmt.Errorf("failed to run container: %v", err)
+	}
+	if len(result.Warnings) != 0 {
+		fmt.Printf("WARNING: %s \n", strings.Join(result.Warnings, "\n"))
+	}
+
+	// pouch run not specify --name
+	if containerName == "" {
+		containerName = result.Name
+	}
+
+	if rc.rm {
+		defer func() {
+			if err := apiClient.ContainerRemove(containerName, false); err != nil {
+				fmt.Printf("%v\n", err)
+			}
+		}()
+	}
+
+	var wait chan struct{}
+	if rc.attach || rc.stdin {
+		in, out, err := setRawMode(rc.stdin, false)
+		if err != nil {
+			return fmt.Errorf("failed to set raw mode")
+		}
+		defer func() {
+			if err := restoreMode(in, out); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to restore term mode")
+			}
+		}()
+
+		conn, br, err := apiClient.ContainerAttach(containerName, rc.stdin)
+		if err != nil {
+			return fmt.Errorf("failed to attach container: %v", err)
+		}
+
+		wait = make(chan struct{})
+		go func() {
+			io.Copy(os.Stdout, br)
+			close(wait)
+		}()
+		go func() {
+			io.Copy(conn, os.Stdin)
+			close(wait)
```

Comment:
It maybe panic if tries to close closed channel.

Reply:
Yes, you are right, i will fix it, @0x04C2